### PR TITLE
Ruby: add support to circuit_breaking xds interop test case

### PR DESCRIPTION
**Pull request by stanley-cheung**
b/173545392

Add support to Ruby for the `circuit_breaking` xDS interop test case
 
Key changes:
- The ruby test client now responds to the `Configure()` RPCs to change the list of RPCs to send and their corresponding metadata
- The test client now responds to `GetClientAccumulatedStats()` RPCs to return additional accumulated view of the # rpcs started/succeeded/failed
- In order to accommodate the `rpc-behavior: keep-open` behavior expected from the test runner, the test client will start an RPC in a thread instead of in a loop in order to not block the test client. This will allow us to keep hundreds of RPCs open and unresolved to test out the circuit breaking feature.

Here's one adhoc run of the `grpc_xds_ruby` job that has all 12 test cases including `circuit_breaking` passing: https://g3c.corp.google.com/results/invocations/e2a2647a-a0e2-4da8-ab6a-375c58bd9a7b/targets

There was a slight inconsistency that the existing `GetClientStats()` is expecting the RPC name to be key'ed as `UnaryCall` and `EmptyCall`, but the new `GetClientAccumulatedStats()` is expecting `UNARY_CALL` and `EMPTY_CALL`, because the former was coming as the `--rpc` parameter as a string, and the latter is coming from a proto message as an enum. We can address this at a later time. In the meantime, there was some additional conversion that needs to be done to make sure we are dealing with the right key.

### CODE REVIEW

File: tools/run_tests/run_xds_tests.py [hunk 1]
```diff
@@ -1282,6 +1282,9 @@ def test_circuit_breaking(gcp, original_backend_service, instance_group,
         logger.info('UNARY_CALL reached stable state after increase (%d)',
                     extra_backend_service_max_requests)
         logger.info('success')
+        configure_client([
```
Comment (stanley-cheung) [inline tools/run_tests/run_xds_tests.py:1285]: I propose this change just as a nice-to have. This is the end of the `test_circuit_breaking` function in the Python test runner after the test case is successful. I propose we call `Configure()` once more to the test client to reset it to be only sending `UnaryCall` with empty metadata at this stage. The test runner will eventually kill the test client, but in the meantime, since the last `Configure()` call is with `rpc-behavior: keep-open`, in this particular instance the Ruby test client will keep opening new threads to send those RPCs. It takes a few minutes for the cleanup to happen so in those few minutes, the Ruby test client could have easily started tens of thousands of new threads. It did not cause any problem, per se, as the tests will still pass as those RPCs are irrelevant and the test client will be killed eventually. So this is just a nice-to-have.

Comment (voidzcy) [reply tools/run_tests/run_xds_tests.py:1285]: nit: it would be good to add some note. Something like

```
// Avoid new RPCs being outstanding (some test clients create threads
// for sending RPCs) after restoring backend services.
```

File: tools/internal_ci/linux/grpc_xds_ruby_test_in_docker.sh [hunk 2]
```diff
@@ -60,11 +60,11 @@ touch "$TOOLS_DIR"/src/proto/grpc/health/v1/__init__.py
 
 (cd src/ruby && bundle && rake compile)
 
-GRPC_VERBOSITY=debug GRPC_TRACE=xds_client,xds_resolver,xds_cluster_manager_lb,cds_lb,xds_cluster_resolver_lb,priority_lb,xds_cluster_impl_lb,weighted_target_lb "$PYTHON" \
+GRPC_VERBOSITY=debug GRPC_TRACE=xds_client,xds_resolver,xds_cluster_manager_lb,cds_lb,priority_lb,xds_cluster_impl_lb,weighted_target_lb "$PYTHON" \
```
Comment (voidzcy) [inline tools/internal_ci/linux/grpc_xds_ruby_test_in_docker.sh:63]: You may not want to change this line.

File: src/ruby/pb/test/xds_client.rb [hunk 3]
```diff
@@ -71,6 +88,29 @@ def create_stub(opts)
   )
 end
 
+class ConfigureTarget < Grpc::Testing::XdsUpdateClientConfigureService::Service
+  include Grpc::Testing
+
+  def configure(req, _call)
+    $rpcs_to_send = req['types'];
```
Comment (voidzcy) [inline src/ruby/pb/test/xds_client.rb:95]: We'd want updates for `$rpcs_to_send` and `$metadata_to_send` to be _atomic_. That is, the thread running `run_test_loop` should see them consistently, not interleave with this `configure` call (e.g., `$rpcs_to_send` is updated but `$metadata_to_send` is not).

In Java, we created a data structure ([`RpcConfig`](https://github.com/grpc/grpc-java/blob/499694e9da1982f5a23ae6ea8b1fa8b63f63a2c1/interop-testing/src/main/java/io/grpc/testing/integration/XdsTestClient.java#L508)) to hold `rpcs_to_send` and `metadata_to_send`. Each time this configure() RPC is invoked, it creates a new `RpcConfig` instance to replace the global reference.

In C-core, it is similar, but just slightly different representations (It uses [a queue of configs](https://github.com/grpc/grpc/blob/e33849682c410452596ebf008a2b984e388e4f99/test/cpp/interop/xds_interop_client.cc#L411-L419), each of which is the configuration for one RPC type).

File: tools/run_tests/run_xds_tests.py (hunk 1 above)
Comment (stanley-cheung) [reply tools/run_tests/run_xds_tests.py]: Added.

File: tools/internal_ci/linux/grpc_xds_ruby_test_in_docker.sh (hunk 2 above)
Comment (stanley-cheung) [reply tools/internal_ci/linux/grpc_xds_ruby_test_in_docker.sh]: Reverted. For some reason I thought that tag was not valid seeing some errors in the logs, but I no longer see those.

File: src/ruby/pb/test/xds_client.rb (hunk 3 above)
Comment (stanley-cheung) [reply src/ruby/pb/test/xds_client.rb]: Fixed. Thanks for the suggestion. I have a `RpcConfig` object now.

**Review by voidzcy — APPROVED** [review]
LGTM

File: src/ruby/pb/test/xds_client.rb [hunk 4]
```diff
@@ -109,29 +164,62 @@ def get_client_stats(req, _call)
...
+  end
   remote_peer
 end
 
+def execute_rpc_in_thread(op, rpc_stats_key)
+  num_open_threads = $keep_open_threads.size
+  if num_open_threads % 50 == 0
+    GRPC.logger.info("number of keep_open_threads = #{num_open_threads}")
+  end
+  $keep_open_threads << Thread.new {
```
Comment (apolcyn) [inline src/ruby/pb/test/xds_client.rb:203]: does `$keep_open_threads` need to be global?

can we change this method to return the new thread, and have `keep_open_threads` be a local of `run_test_loop`?

That way, it's easier to reason about how this is operation thread safe

**Review by apolcyn — CHANGES_REQUESTED** [review]

File: src/ruby/pb/test/xds_client.rb [hunk 5]
```diff
@@ -109,29 +164,62 @@ def get_client_stats(req, _call)
...
+    begin
+      op.execute
+      # The following should _not_ happen with the current spec
+      # because we are only executing RPCs in a thread if we expect it
+      # to be kept open, or deadline_exceeded, or dropped by the load
+      # balancing policy. These RPCs should not complete successfully.
+      # Doing this for consistency
+      $num_rpcs_succeeded_by_method[rpc_stats_key] += 1
```
[the remark this answers is not in the document]
Comment (stanley-cheung) [reply src/ruby/pb/test/xds_client.rb]: Comments addressed. Thanks for pointing that out! I added a `$accumuted_stats_mu` mutex.

File: src/ruby/pb/test/xds_client.rb (hunk 4 above)
Comment (stanley-cheung) [reply src/ruby/pb/test/xds_client.rb]: Comment addressed. Now `keep_open_threads` is local to `run_test_loop`.

**Review by apolcyn — APPROVED** [review]
